### PR TITLE
applications: serial_lte_modem: Send hexadecimal data by data mode

### PR DESCRIPTION
**Pull request by junqingzou**
Remove SLM datatype in all sending functions.
If no payload is provided, switch to data mode automatically.
Remove detection of hex data in receiving functions.
Add data mode to socket send() and sendto().
Move data mode APIs to at host header.
Add data mode documentation.

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/src/slm_at_tcpip.c [hunk 1]
```diff
@@ -490,6 +478,35 @@ static int do_send(const uint8_t *data, int datalen)
 	}
 }
 
+static int do_send_datamode(const uint8_t *data, int datalen)
+{
+	uint32_t offset = 0;
+	int ret = 0;
+	int sock = client.sock;
+
+	/* For TCP/TLS Server, send to imcoming socket */
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_tcpip.c:487]:
```suggestion
	/* For TCP/TLS Server, send to incoming socket */
```

File: applications/serial_lte_modem/src/slm_at_tcpip.c [hunk 2]
```diff
@@ -490,6 +478,35 @@ static int do_send(const uint8_t *data, int datalen)
...
+			LOG_ERR("No remote connection");
+			return -EINVAL;
+		}
+	}
+
+	while (offset < datalen) {
+		ret = send(sock, data + offset, datalen - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send() failed: %d", -errno);
+			ret = -errno;
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_tcpip.c:501]: `ret` seems to be unused anyway? Should the error code be assigned to `offset` instead?

File: applications/serial_lte_modem/src/slm_at_tcpip.c [hunk 3]
```diff
@@ -609,6 +608,30 @@ static int do_sendto(const char *url, uint16_t port, const uint8_t *data,
...
+	if (ret < 0) {
+		return ret;
+	}
+
+	while (offset < datalen) {
+		ret = sendto(client.sock, data + offset, datalen - offset, 0,
+			(struct sockaddr *)&remote, sizeof(struct sockaddr_in));
+		if (ret <= 0) {
+			LOG_ERR("sendto() failed: %d", -errno);
+			ret = -errno;
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_tcpip.c:626]: Same here, `ret` seems to be unused.

**Review by rlubos — COMMENTED** [review]
Looks good, I didn't catch anything significant. This seems to me however like a significant change, I think it should get an entry in the release notes.

**Review by FrancescoSer — APPROVED** [review]

**Comment by junqingzou** [comment]
> Looks good, I didn't catch anything significant. This seems to me however like a significant change, I think it should get an entry in the release notes.

Added

File: applications/serial_lte_modem/src/slm_at_tcpip.c (hunk 2 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_tcpip.c:501]: Used as return value in the end

File: applications/serial_lte_modem/src/slm_at_tcpip.c (hunk 3 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_tcpip.c:626]: Used as return value in the end

File: applications/serial_lte_modem/src/slm_at_tcpip.c (hunk 2 above)
Comment (rlubos) [reply applications/serial_lte_modem/src/slm_at_tcpip.c:550]: Either I miss something, or there is no change.

Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_tcpip.c:550]: Sorry, it's copied from do_send() and indeed not used in this function, will be removed

**Review by rlubos — APPROVED** [review]

**Review by lats1980 — APPROVED** [review]
LGTM. I suggest to add datamode example into document. Please add such example in the future work.

**Comment by junqingzou** [comment]
> LGTM. I suggest to add datamode example into document. Please add such example in the future work.

Yes will add more in future PRs.
The data mode page only explains mechanism but does lack of samples.

**Pull request merged**